Add docs page explaining Gloo resources storage (#406)
solo-io/gloo: docs/operator_guide/understanding_gloo_configuration.md

# Understanding Gloo configuration

---
**NOTE**

> Gloo currently requires a running Kubernetes cluster to use as data store. We are working on supporting additional storage options.

Reviewer: shoul
```suggestion
Gloo currently requires a running Kubernetes cluster to use as data store. We are adding additional storage options in an upcoming release. For a quick test run of gloo, you can deploy gloo on minikube.
```

[custom resources](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources).
Each of the Gloo configuration objects (virtual services, upstreams, etc.) is an instance of a 
[Kubernetes Custom Resource Definition (CRD)](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions).

Storing configuration as Kubernetes CRDs has several advantages:
1. **All Gloo resources can be managed using standard Kubernetes APIs and tools.** You can interact with them using
`kubectl` just as you would with normal Kubernetes resources. Moreover, you can perform the same actions programmatically 
via the [Kubernetes REST API](https://kubernetes.io/docs/reference/using-api/api-overview/) and the provided 
[Client Libraries](https://kubernetes.io/docs/reference/using-api/client-libraries/) built on top of it.
2. **Your Gloo configuration lives close to the resources it applies to.** You don't need to run any additional software. 
Your services and the configurations that define how traffic is routed to them share the same infrastructure and APIs. 

> Yet, by being defined as different resources, possibly in different namespaces, they are completely decoupled and 

Reviewer: ```suggestion
Since Gloo resources are defined as their own resource class, they are completely decoupled and isolated from one another.
```

> isolated from each other. You dont't have to worry about changes in your Kubernetes resources affecting your Gloo CRDs.

Reviewer: ```suggestion
isolated from each other. You don't have to worry about changes in your Kubernetes resources affecting your Gloo CRDs.
```

    - '*'
    name: gloo-system.default
    routes:
    - matcher:
        exact: /sample-route-1
      routeAction:
        single:
          upstream:
            name: default-petstore-8080
            namespace: gloo-system
      routePlugins:
        prefixRewrite:
          prefixRewrite: /api/pets
status:
  reported_by: gateway
  state: 1
  subresource_statuses:
    '*v1.Proxy gloo-system gateway-proxy':
      reported_by: gloo
      state: 1
```

Note how the above document is a superset of the one returned by `glooctl`.

We are able to access _Virtual Services_ using `kubectl` because Gloo registered a `VirtualService` CRD with the Kubernetes 
API server. You can verify this by running `kubectl get crds`, which should return a list similar to this one:

```bash
NAME                              CREATED AT
gateways.gateway.solo.io          2019-02-11T20:37:04Z
proxies.gloo.solo.io              2019-02-11T20:37:04Z
settings.gloo.solo.io             2019-02-11T20:36:55Z
upstreams.gloo.solo.io            2019-02-11T20:37:01Z
virtualservices.gateway.solo.io   2019-02-11T20:37:04Z
```

You can use the short resource name instead of the fully qualified one in your commands 

> (e.g. `kubectl get vs -n gloo-system` instead of `kubectl get virtualservices.gateway.solo.io/default -n gloo-system`). 

Reviewer: ```suggestion
(e.g. `kubectl get virtualservice -n gloo-system` or even `kubectl get vs -n gloo-system` instead of `kubectl get virtualservices.gateway.solo.io/default -n gloo-system`). 
```
(vs is a bit of a leap of faith)

> Try to run `kc describe crds/virtualservices.gateway.solo.io` to see what a Custom Resource Definition for a Gloo 

Reviewer: ```suggestion
Try to run `kubectl describe crds/virtualservices.gateway.solo.io` to see what a Custom Resource Definition for a Gloo 
```

rules:
- apiGroups: ["gateway.solo.io/v1"]
  resources: ["virtualservices"]
  verbs: ["get", "watch", "list"]
```

```yaml
kind: RoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: read-virtualservices
  namespace: gloo-system
subjects:
- kind: Group
  name: Developers
  apiGroup: rbac.authorization.k8s.io
roleRef:
  kind: Role
  name: virtualservice-reader
  apiGroup: rbac.authorization.k8s.io
```

```yaml
kind: Role
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  namespace: gloo-system
  name: virtualservice-admin
rules:
- apiGroups: ["gateway.solo.io/v1"]
  resources: ["virtualservices"]
  verbs: ["get", "watch", "list", "create", "update", "patch", "delete"]
```

```yaml
kind: RoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: admin-virtualservices
  namespace: gloo-system
subjects:
- kind: User
  name: Alice
  apiGroup: rbac.authorization.k8s.io
roleRef:
  kind: Role
  name: virtualservice-admin
  apiGroup: rbac.authorization.k8s.io
```

After these resources have been created, if `Bob`, who belongs to the `Developers` group, tries to create a virtual 
service by running

```bash
kubectl apply -f ~/my-virtual-service.yaml
```

> she will be presented with an error message:

Reviewer: ```suggestion
he will be presented with an error message:
```
(referring to Bob, I'm fine either way, I just think this is less distracting and possibly what you meant)

Reviewer: looks great, just a few tweak suggestions

# Understanding Gloo configuration

---
**NOTE**

Gloo currently requires a running Kubernetes cluster to use as data store. We are adding additional storage options in 
an upcoming release. For a quick test run of Gloo, you can deploy Gloo on `minikube`.

To be notified of the most recent updates, follow us on [Twitter](https://twitter.com/soloio_inc) and join our 
[community Slack channel](http://slack.solo.io/).

---

## Configuration storage
By default, Gloo leverages Kubernetes to implement its 
[declarative infrastructure model](./gitops/gloo_declarative_model.md#gloo-as-declarative-infrastructure).
The Gloo configuration consists of a set of YAML documents that are stored in Kubernetes as
[custom resources](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources).

> Each of the Gloo configuration objects (virtual services, upstreams, etc.) is an instance of a 

Author: I am using "configuration" in a generic way here, meaning all the gloo CRDs. That would include secrets, gateway, etc., making the list a bit long. Does that make sense?

> Each of the Gloo configuration objects (virtual services, upstreams, etc.) is an instance of a 

Reviewer 2: sure
